R/gh_response.R

```diff
@@ -14,20 +16,52 @@ gh_process_response <- function(response) {
     res <- list(message = content(response, as = "text"))
   }
 
-  if (status_code(response) >= 300) {
-    cond <- structure(list(
-      call = sys.call(-1),
-      content = res,
-      headers = heads,
-      message = paste0("GitHub API error (", status_code(response), "): ",
-                       heads$`status`, "\n  ", res$message, "\n")
-    ), class = c("github_error", paste0("http_error_", status_code(response)), "error", "condition"))
-    stop(cond)
-  }
-
   attr(res, "method") <- response$request$method
-  attr(res, "response") <- heads
+  attr(res, "response") <- headers(response)
   attr(res, ".send_headers") <- response$request$headers
   class(res) <- c("gh_response", "list")
   res
 }
+
+## https://developer.github.com/v3/#client-errors
+gh_error <- function(response, call = sys.call(-1)) {
+  heads <- headers(response)
+  res <- content(response)
+  status <- status_code(response)
+
+  msg <- c(
+    "",
+    paste0("GitHub API error (", status, "): ", heads$status),
+    paste0("Message: ", res$message)
+  )
+
+  doc_url <- res$documentation_url
+  if (!is.null(doc_url)) {
+    msg <- append(msg, paste0("Read more at ", doc_url))
+  }
+
+  errors <- res$errors
+  if (!is.null(errors)) {
+    errors <- as.data.frame(do.call(rbind, errors))
+    nms <- c("resource", "field", "code", "message")
+    nms <- nms[nms %in% names(errors)]
+    msg <- append(
+      msg,
+      c("",
+        "Errors:",
+        capture.output(print(errors[nms], row.names = FALSE))
```

Comment:
Is it possible to avoid `capture.output` somehow? It does not play well with debugging, and terminal coloring, so I think it is good practice to avoid it, if possible.

Reply:
It seems difficult to avoid it. There is `format.data.frame()`, but that still returns a data frame.

Reply:
Yes, I tried `format.data.frame()` myself, among other things. I don't know of any other way to get this result for so little work. If it really matters, I can do this another way.

Reply:
I think it would be better to avoid constructing a `data.frame` at all in this case and just use `print(sep = " ", collapse = "\n")` to construct the format directly.

Reply:
Although I just realized this is already merged, so I guess I am too slow!

Reply:
We can still change it, if you think it is better. I really don't like `capture.output`. Although in this case it is mostly harmless.

Reply:
It's a PITA to format nicely because the number and names of the fields varies across endpoints (see my two examples).

Reply:
Yeah, I think it is just fine as it is. At some point we might want to write some real `format.data.frame()` method, though. That returns actual strings.

Reply:
BTW I think it's good to keep capturing examples of error-triggering calls from real life, so we can sanity check this function as it evolves. It would be impossible to write a useful version of this function based only on GitHub's API docs.